### ISSUE TITLE
Fail to produce a validation message on an array. validate_config_against_schema gets a TypeError Exception

### ISSUE THREAD

**Issue by NikosVlagoidis** [StackStorm/st2#4166]
##### ISSUE TYPE

 - Bug Report *
 - Feature Idea
 - Other

##### STACKSTORM VERSION

```
st2 2.8dev (0f1d9dc), on Python 2.7.6
```

##### OS / ENVIRONMENT / INSTALL METHOD

docker
##### SUMMARY

There is a problem when validating an array of objects. If the validation fails then on this line:

[https://github.com/StackStorm/st2/blob/master/st2common/st2common/util/pack.py#L112
](url)

Because the path has a number you cannot join an integer.

```
Traceback (most recent call last):
  File "/opt/stackstorm/st2/local/lib/python2.7/site-packages/st2common/router.py", line 472, in __call__
    resp = func(**kw)
  File "/opt/stackstorm/st2/local/lib/python2.7/site-packages/st2api/controllers/v1/pack_configs.py", line 107, in put
    config_api.validate(validate_against_schema=True)
  File "/opt/stackstorm/st2/local/lib/python2.7/site-packages/st2common/models/api/pack.py", line 274, in validate
    cleaned_values = self._validate_config_values_against_schema()
  File "/opt/stackstorm/st2/local/lib/python2.7/site-packages/st2common/models/api/pack.py", line 296, in _validate_config_values_against_schema
    pack_name=self.pack)
  File "/opt/stackstorm/st2/local/lib/python2.7/site-packages/st2common/util/pack.py", line 111, in validate_config_against_schema
    attribute = '.'.join(attribute)
TypeError: sequence item 1: expected string or Unicode, int found
2018-06-06 14:24:35,673 139995931735056 ERROR error_handling [-] API call failed: sequence item 1: expected string or Unicode, int found
Traceback (most recent call last):

```

When I changed the code to:

```
attribute = '.'.join(str(attribute))

```

MESSAGE: Failed validating attribute "d.e.q.u.e.(.[.u.'.i.n.s.t.a.n.c.e.s.'.,. .0.,. .u.'.a.l.i.a.s.'.].)" in config for pack "my_pack" (/opt/stackstorm/configs/my_pack.yaml): u'alias' is a required property.

I guess there is a problem with path when you have an array becuase you get for example 
instances[0]['alias'] for a path and the exception cannot produce a message

**Comment by Kami** [comment]
Could it be related to changes in this PR - https://github.com/StackStorm/st2/pull/4122?

@NikosVlagoidis Please also post `config.schema.yaml` and config file for pack in question which reproduces this problem.

**Comment by NikosVlagoidis** [comment]
config.schema.yaml
```yaml
---
  instances:
    type: "array"
    required: false
    items:
      type: "object"
      properties:
          uid:
              description: "a unique id for this instance"
              type: "string"
              required: false
              secret: false
          alias:
              description: "an alias for this instance"
              type: "string"
              required: true
              secret: false
          base_url:
              description: "Base Url"
              type: "string"
              secret: false
              required: true
          token:
              description: "token"
              type: "string"
              secret: true
              required: true
```

The error occured when I tried to update the config throughh the Python st2 client witch uses st2 api.

The error log is from st2api.log

**Comment by Kami** [comment]
What about the config you are using (you can mask private / sensitive / secret values)?

Also, until we dig in, a work around might be to utilize an object instead of an array.

We do that for in many packs, an example can be found in Liblcoud pack - https://github.com/stackstorm-exchange/stackstorm-libcloud.

**Comment by Kami** [comment]
Bug fix for this issue is in #4168.

Thanks again for reporting this and sorry for the confusion.

The original issue description didn't contain enough context and details so I was a bit confused and wasn't totally sure where the problem lies.

I assumed the problem lies elsewhere, but it was just an issue of throwing an unrelated, non-user friendly exception if an array config item failed config validation. So the actual functionality worked fine, the issue is just related to the error we throw.

**Comment by NikosVlagoidis** [comment]
thank you! that was the issue.
